File: apps/web/src/stores/cookies.ts
import type { ResponseCookie } from "@rabbitpost/shared";
import { create } from "zustand";
import { persist } from "zustand/middleware";

/** Cookie Jar 中的单条 cookie：以 Set-Cookie 风格原始串存储（同 Postman 编辑形态） */
export interface JarCookie {
  id: string;
  /** 形如 "Cookie_1=value; Path=/; Expires=Tue, 19 Jan 2038 03:14:07 GMT;" */
  raw: string;
}

/** 按域名分组的 cookie 集合 */
export interface JarDomain {
  domain: string;
  cookies: JarCookie[];
}

/** raw cookie 串解析结果（属性名大小写不敏感） */
export interface ParsedCookie {
  name: string;
  value: string;
  path: string;
  domain?: string;
  expires?: string;
  secure: boolean;
  httpOnly: boolean;
}

let seq = 0;
function newCookieId(): string {
  return `ck-${Date.now()}-${seq++}`;
}

/** 解析 "name=value; Attr=..." 形式的 cookie 串；无 name 时返回 null */
export function parseRawCookie(raw: string): ParsedCookie | null {
  const segments = raw
    .split(";")
    .map((s) => s.trim())
    .filter(Boolean);
  const first = segments.shift();
  if (!first) return null;
  const eq = first.indexOf("=");
  const name = (eq >= 0 ? first.slice(0, eq) : first).trim();
  if (!name) return null;
  const cookie: ParsedCookie = {
    name,
    value: eq >= 0 ? first.slice(eq + 1).trim() : "",
    path: "/",
    secure: false,
    httpOnly: false,
  };
  for (const seg of segments) {
    const i = seg.indexOf("=");
    const key = (i >= 0 ? seg.slice(0, i) : seg).trim().toLowerCase();
    const val = i >= 0 ? seg.slice(i + 1).trim() : "";
    if (key === "path" && val) cookie.path = val;
    else if (key === "domain" && val) cookie.domain = val.replace(/^\./, "");
    else if (key === "expires") cookie.expires = val;
    else if (key === "max-age" && val) {
      const n = Number(val);
      if (!Number.isNaN(n)) cookie.expires = new Date(Date.now() + n * 1000).toUTCString();
    } else if (key === "secure") cookie.secure = true;
    else if (key === "httponly") cookie.httpOnly = true;
  }
  return cookie;
}

/** 规范化用户输入的域名：去掉协议 / 路径 / 端口 / 前导点，转小写 */
export function normalizeDomainInput(input: string): string {
  return input
    .trim()
    .toLowerCase()
    .replace(/^[a-z]+:\/\//, "")
    .replace(/\/.*$/, "")
    .replace(/:\d+$/, "")
    .replace(/^\./, "");
}

/** cookie 是否已过期（Expires 可解析且早于当前时间） */
function isExpired(expires: string | undefined): boolean {
  if (!expires) return false;
  const t = Date.parse(expires);
  return !Number.isNaN(t) && t < Date.now();
}

/**
 * 依据 URL 从 Cookie Jar 中筛出可发送的 cookie，拼为 Cookie 请求头值。
 * 匹配规则同浏览器：域名后缀匹配、路径前缀匹配、Secure 仅 https、过期跳过。
 */
export function cookieHeaderForUrl(rawUrl: string, domains: JarDomain[]): string {
  let url: URL;
  try {
    // 与服务端 executor 一致：缺协议时补 http://
    url = new URL(/^https?:\/\//i.test(rawUrl) ? rawUrl : `http://${rawUrl}`);
  } catch {
    return "";
  }
  const host = url.hostname.toLowerCase();
  const pairs: string[] = [];
  const seen = new Set<string>();
  for (const d of domains) {
    for (const c of d.cookies) {
      const parsed = parseRawCookie(c.raw);
      if (!parsed) continue;
      const domain = (parsed.domain ?? d.domain).toLowerCase();
      if (host !== domain && !host.endsWith(`.${domain}`)) continue;
      if (!url.pathname.startsWith(parsed.path)) continue;
      if (parsed.secure && url.protocol !== "https:") continue;
      if (isExpired(parsed.expires)) continue;
      if (seen.has(parsed.name)) continue;
      seen.add(parsed.name);
      pairs.push(`${parsed.name}=${parsed.value}`);
    }
  }
  return pairs.join("; ");
}

/** 从 URL 提取 hostname；非法 URL 返回 null */
export function hostnameOf(rawUrl: string): string | null {
  try {
    return new URL(/^https?:\/\//i.test(rawUrl) ? rawUrl : `http://${rawUrl}`).hostname.toLowerCase();
  } catch {
    return null;
  }
}

/** 将响应解析出的 ResponseCookie 还原为 raw 串 */
function buildRaw(c: ResponseCookie): string {
  const parts = [`${c.name}=${c.value}`];
  parts.push(`Path=${c.path ?? "/"}`);
  if (c.domain) parts.push(`Domain=${c.domain.replace(/^\./, "")}`);
  if (c.expires) parts.push(`Expires=${c.expires}`);
  else if (c.maxAge !== undefined)
    parts.push(`Expires=${new Date(Date.now() + c.maxAge * 1000).toUTCString()}`);
  if (c.secure) parts.push("Secure");
  if (c.httpOnly) parts.push("HttpOnly");
  return `${parts.join("; ")};`;
}

interface CookiesState {
  domains: JarDomain[];
  /** 允许脚本编程访问 cookie 的域名白名单（同 Postman Domains allowlist） */
  allowlist: string[];

  addDomain: (domain: string) => void;
  removeDomain: (domain: string) => void;
  /** id 为 null 时新增，否则更新对应 cookie 的 raw 串 */
  upsertCookie: (domain: string, id: string | null, raw: string) => void;
  removeCookie: (domain: string, id: string) => void;
  clearAll: () => void;
  addAllowlistDomain: (domain: string) => void;
  removeAllowlistDomain: (domain: string) => void;
  /** 将响应 Set-Cookie 写入 Jar：同名覆盖，已过期则删除（同浏览器语义） */
  storeResponseCookies: (host: string, cookies: ResponseCookie[]) => void;
}

export const useCookiesStore = create<CookiesState>()(
  persist(
    (set) => ({
      domains: [],
      allowlist: [],

      addDomain: (domain) =>
        set((s) =>
          s.domains.some((d) => d.domain === domain)
            ? s
            : { domains: [...s.domains, { domain, cookies: [] }] },
        ),

      removeDomain: (domain) =>
        set((s) => ({ domains: s.domains.filter((d) => d.domain !== domain) })),

      upsertCookie: (domain, id, raw) =>
        set((s) => ({
          domains: s.domains.map((d) =>
            d.domain !== domain
              ? d
              : {
                  ...d,
                  cookies: id
                    ? d.cookies.map((c) => (c.id === id ? { ...c, raw } : c))
                    : [...d.cookies, { id: newCookieId(), raw }],
                },
          ),
        })),

      removeCookie: (domain, id) =>
        set((s) => ({
          domains: s.domains.map((d) =>
            d.domain !== domain
              ? d
              : { ...d, cookies: d.cookies.filter((c) => c.id !== id) },
          ),
        })),

      clearAll: () => set({ domains: [] }),

      addAllowlistDomain: (domain) =>
        set((s) =>
          s.allowlist.includes(domain) ? s : { allowlist: [...s.allowlist, domain] },
        ),

      removeAllowlistDomain: (domain) =>
        set((s) => ({ allowlist: s.allowlist.filter((d) => d !== domain) })),

      storeResponseCookies: (host, cookies) =>
        set((s) => {
          let domains = [...s.domains];
          for (const c of cookies) {
            if (!c.name) continue;
            const domain = (c.domain ?? host).replace(/^\./, "").toLowerCase();
            const expired =
              (c.maxAge !== undefined && c.maxAge <= 0) || isExpired(c.expires);
            let entry = domains.find((d) => d.domain === domain);
            if (!entry) {
              if (expired) continue;
              entry = { domain, cookies: [] };
              domains.push(entry);
            }
            // 同名 cookie 覆盖；过期视为删除
            const rest = entry.cookies.filter(
              (k) => parseRawCookie(k.raw)?.name !== c.name,
            );
            const next = expired ? rest : [...rest, { id: newCookieId(), raw: buildRaw(c) }];
            domains = domains.map((d) =>
              d.domain === domain ? { ...d, cookies: next } : d,
            );
          }
          return { domains };
        }),
    }),
    { name: "rp.cookies" },
  ),
);

File: crates/rp-core/src/cookies.rs
//! 跨请求 Cookie Jar（对齐 newman --cookie-jar / --export-cookie-jar）。
//! 引擎不启用 reqwest 的自动 cookie 管理（避免与 Runner 长任务串味），
//! 由调用方持有 Jar 并显式传入执行上下文；CLI 一次 run 内共享一个 Jar。
//!
//! 文件格式兼容两种：Postman 导出的 cookie jar（外层数组或 { cookies: [...] }，
//! 元素带 name/value/domain/path/secure/expires）与极简的 [{name,value,domain,...}]。
use std::sync::Mutex;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Cookie {
    pub name: String,
    #[serde(default)]
    pub value: String,
    /// 前导点已剥除；host_only 表示只允许精确域名匹配
    pub domain: String,
    #[serde(default = "default_path")]
    pub path: String,
    #[serde(default)]
    pub secure: bool,
    /// 过期时间（Unix 秒）；None 为会话 cookie。容忍 Postman 导出的字符串形态
    #[serde(
        default,
        deserialize_with = "de_expires",
        skip_serializing_if = "Option::is_none"
    )]
    pub expires: Option<i64>,
    #[serde(default, rename = "hostOnly", skip_serializing_if = "is_false")]
    pub host_only: bool,
    /// Postman 导出里的 httpOnly 标记，序列化保留
    #[serde(default, rename = "httpOnly", skip_serializing_if = "is_false")]
    pub http_only: bool,
}

fn default_path() -> String {
    "/".to_string()
}

fn is_false(v: &bool) -> bool {
    !*v
}

/// expires 兼容数字与字符串（Postman 导出两种都出现过）
fn de_expires<'de, D>(deserializer: D) -> Result<Option<i64>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let value = Option::<serde_json::Value>::deserialize(deserializer)?;
    Ok(match value {
        Some(serde_json::Value::Number(n)) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)),
        Some(serde_json::Value::String(s)) => s.trim().parse::<i64>().ok(),
        _ => None,
    })
}

/// 线程安全的 Cookie 存储：同一 run 内所有请求共享
#[derive(Debug, Default)]
pub struct CookieJar {
    cookies: Mutex<Vec<Cookie>>,
}

impl CookieJar {
    pub fn new() -> Self {
        Self::default()
    }

    /// 从 JSON 文本加载（Postman cookie jar 导出 / 极简数组）
    pub fn load_json(text: &str) -> anyhow::Result<Self> {
        let raw: serde_json::Value = serde_json::from_str(text)
            .map_err(|e| anyhow::anyhow!("cookie jar JSON 解析失败：{e}"))?;
        let list = match &raw {
            serde_json::Value::Array(_) => &raw,
            serde_json::Value::Object(_) => raw
                .get("cookies")
                .ok_or_else(|| anyhow::anyhow!("cookie jar 对象须包含 cookies 数组"))?,
            _ => anyhow::bail!("cookie jar 须为数组或含 cookies 字段的对象"),
        };
        let mut cookies: Vec<Cookie> = serde_json::from_value(list.clone())?;
        for cookie in &mut cookies {
            cookie.domain = cookie.domain.trim_start_matches('.').to_string();
            if cookie.path.is_empty() {
                cookie.path = default_path();
            }
            // Postman 的 expires 可能是字符串
        }
        Ok(Self {
            cookies: Mutex::new(cookies),
        })
    }

    /// 导出为 JSON（极简数组，Postman 兼容字段名）
    pub fn to_json(&self) -> String {
        let cookies = self.cookies.lock().unwrap();
        serde_json::to_string_pretty(&*cookies).unwrap_or_else(|_| "[]".to_string())
    }

    pub fn len(&self) -> usize {
        self.cookies.lock().unwrap().len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// 计算发给 url 的 Cookie 头值；无匹配返回 None
    pub fn cookies_for(&self, url: &str) -> Option<String> {
        let parsed = url::Url::parse(url).ok()?;
        let host = parsed.host_str()?.to_lowercase();
        let path = parsed.path();
        let is_https = parsed.scheme() == "https";
        let now = now_epoch();

        let mut cookies = self.cookies.lock().unwrap();
        // 顺手清理过期 cookie
        cookies.retain(|c| c.expires.is_none_or(|exp| exp > now));
        let pairs: Vec<String> = cookies
            .iter()
            .filter(|c| {
                if c.secure && !is_https {
                    return false;
                }
                let domain_match = if c.host_only {
                    host == c.domain
                } else {
                    host == c.domain || host.ends_with(&format!(".{}", c.domain))
                };
                domain_match && path.starts_with(&c.path)
            })
            .map(|c| format!("{}={}", c.name, c.value))
            .collect();
        if pairs.is_empty() {
            None
        } else {
            Some(pairs.join("; "))
        }
    }

    /// 解析一条 Set-Cookie 头值并入库（同名同域同路径覆盖）
    pub fn store(&self, url: &str, set_cookie: &str) {
        let Some(parsed_url) = url::Url::parse(url).ok() else {
            return;
        };
        let host = parsed_url.host_str().unwrap_or_default().to_lowercase();
        if host.is_empty() {
            return;
        }

        let mut parts = set_cookie.split(';');
        let Some(pair) = parts.next() else { return };
        let Some((name, value)) = pair.split_once('=') else { return };
        let name = name.trim().to_string();
        if name.is_empty() {
            return;
        }
        let value = value.trim().to_string();

        let mut cookie = Cookie {
            name,
            value,
            domain: host.clone(),
            path: default_path(),
            secure: false,
            expires: None,
            host_only: true,
            http_only: false,
        };
        for attr in parts {
            let attr = attr.trim();
            let (key, val) = attr
                .split_once('=')
                .map(|(k, v)| (k.trim(), v.trim()))
                .unwrap_or((attr, ""));
            match key.to_ascii_lowercase().as_str() {
                "domain" => {
                    cookie.domain = val.trim_start_matches('.').to_lowercase();
                    cookie.host_only = false;
                }
                "path" => {
                    if !val.is_empty() {
                        cookie.path = val.to_string();
                    }
                }
                "secure" => cookie.secure = true,
                "httponly" => cookie.http_only = true,
                "max-age" => {
                    if let Ok(secs) = val.parse::<i64>() {
                        cookie.expires = Some(now_epoch() + secs);
                    }
                }
                "expires" => {
                    if cookie.expires.is_none() {
                        cookie.expires = parse_http_date(val);
                    }
                }
                _ => {}
            }
        }

        let mut cookies = self.cookies.lock().unwrap();
        // Max-Age=0 / 已过期：删除而非存储
        if cookie.expires.is_some_and(|exp| exp <= now_epoch()) {
            cookies.retain(|c| {
                !(c.name == cookie.name && c.domain == cookie.domain && c.path == cookie.path)
            });
            return;
        }
        match cookies
            .iter_mut()
            .find(|c| c.name == cookie.name && c.domain == cookie.domain && c.path == cookie.path)
        {
            Some(existing) => *existing = cookie,
            None => cookies.push(cookie),
        }
    }
}

fn now_epoch() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

/// 解析 HTTP-date（IMF-fixdate，如 "Wed, 21 Oct 2015 07:28:00 GMT"）。
/// 只支持标准 GMT 形态，解析失败按会话 cookie 处理。
fn parse_http_date(value: &str) -> Option<i64> {
    let value = value.trim();
    // 去掉星期前缀 "Wed, "
    let rest = value.split_once(", ").map(|(_, r)| r).unwrap_or(value);
    // 21 Oct 2015 07:28:00 GMT
    let mut parts = rest.split_whitespace();
    let day: i64 = parts.next()?.parse().ok()?;
    let month = match parts.next()? {
        "Jan" => 1,
        "Feb" => 2,
        "Mar" => 3,
        "Apr" => 4,
        "May" => 5,
        "Jun" => 6,
        "Jul" => 7,
        "Aug" => 8,
        "Sep" => 9,
        "Oct" => 10,
        "Nov" => 11,
        "Dec" => 12,
        _ => return None,
    };
    let year: i64 = parts.next()?.parse().ok()?;
    let time = parts.next()?;
    let mut segs = time.split(':');
    let hour: i64 = segs.next()?.parse().ok()?;
    let min: i64 = segs.next()?.parse().ok()?;
    let sec: i64 = segs.next()?.parse().ok()?;
    Some(days_from_civil(year, month, day) * 86400 + hour * 3600 + min * 60 + sec)
}

/// Howard Hinnant 的 civil->days 算法，避免引入 chrono 依赖
fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = if y >= 0 { y } else { y - 399 } / 400;
    let yoe = y - era * 400;
    let mp = (m + 9) % 12;
    let doy = (153 * mp + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146097 + doe - 719468
}

// ---------------------------------------------------------------------------
// 单元测试
// ---------------------------------------------------------------------------
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn store_and_match_basic() {
        let jar = CookieJar::new();
        jar.store("https://api.example.com/users", "sid=abc; Path=/");
        assert_eq!(
            jar.cookies_for("https://api.example.com/users/1").as_deref(),
            Some("sid=abc")
        );
        // hostOnly（未带 Domain 属性）：子域不匹配
        assert!(jar.cookies_for("https://sub.api.example.com/").is_none());
        // 其它域不匹配
        assert!(jar.cookies_for("https://other.com/").is_none());
    }

    #[test]
    fn domain_attr_enables_subdomain_and_secure_gates_http() {
        let jar = CookieJar::new();
        jar.store(
            "https://example.com/",
            "t=1; Domain=.example.com; Secure; Path=/api",
        );
        assert_eq!(
            jar.cookies_for("https://sub.example.com/api/x").as_deref(),
            Some("t=1")
        );
        // 路径前缀不匹配
        assert!(jar.cookies_for("https://example.com/web").is_none());
        // secure cookie 不发 http
        assert!(jar.cookies_for("http://example.com/api").is_none());
    }

    #[test]
    fn overwrite_and_delete() {
        let jar = CookieJar::new();
        jar.store("https://a.com/", "k=1; Path=/");
        jar.store("https://a.com/", "k=2; Path=/");
        assert_eq!(jar.len(), 1);
        assert_eq!(jar.cookies_for("https://a.com/").as_deref(), Some("k=2"));
        // Max-Age=0 删除
        jar.store("https://a.com/", "k=; Path=/; Max-Age=0");
        assert!(jar.is_empty());
    }

    #[test]
    fn json_roundtrip_and_postman_shape() {
        let jar = CookieJar::new();
        jar.store("https://a.com/", "k=v; Domain=a.com; HttpOnly");
        let text = jar.to_json();
        let loaded = CookieJar::load_json(&text).unwrap();
        assert_eq!(loaded.cookies_for("https://a.com/").as_deref(), Some("k=v"));

        // Postman 导出形态：{ "cookies": [...] }，domain 带前导点
        let pm = r#"{"cookies": [{"name": "s", "value": "1", "domain": ".a.com", "path": "/"}]}"#;
        let loaded = CookieJar::load_json(pm).unwrap();
        assert_eq!(
            loaded.cookies_for("https://sub.a.com/").as_deref(),
            Some("s=1")
        );
    }

    #[test]
    fn http_date_parsing() {
        let ts = parse_http_date("Wed, 21 Oct 2015 07:28:00 GMT").unwrap();
        assert_eq!(ts, 1445412480);
        assert!(parse_http_date("garbage").is_none());
    }
}
